02_knight_and_princess/tevfik_aksoy/main.cpp

```diff
@@ -1,93 +1,186 @@
+  Edge(VertexId _source, VertexId _destination, VertexId _id, Color _color)
+      : id(_id), color(_color), source(_source), destination(_destination) {}
+
+  std::string color_to_string(const Edge::Color& color) const;
 
   std::string to_JSON() const {
     std::string json_string;
-    json_string += "\t{ \"id\": " + to_string(id) + ", \"vertex_ids\": [" +
-                   to_string(source) + ", " + to_string(destination) + "] }";
+    json_string += "\t{ \"id\": " + std::to_string(id) + ", \"vertex_ids\": [" +
+                   std::to_string(source) + ", " + std::to_string(destination) +
+                   "], \"color\": \"" + color_to_string(color) + "\" }";
     return json_string;
   }
 };
 
+std::string Edge::color_to_string(const Edge::Color& _color) const {
+  switch (_color) {
+    case Edge::Color::GRAY:
+      return "gray";
+    case Edge::Color::GREEN:
+      return "green";
+    case Edge::Color::BLUE:
+      return "blue";
+    case Edge::Color::YELLOW:
+      return "yellow";
+    case Edge::Color::RED:
+      return "red";
+  }
+}
+
+bool validation::vertex_exists_in_graph(const VertexId& id,
+                                        const vector<Vertex>& vertices) {
+  for (const auto& vertex : vertices)
+    if (vertex.id == id) {
+      return true;
+    }
+  return false;
+}
+bool validation::edge_exists_in_graph(const EdgeId& id,
+                                      const vector<Edge>& edges) {
+  for (const auto& edge : edges)
+    if (edge.id == id) {
+      return true;
+    }
+  return false;
+}
 class Graph {
  public:
-  void insert_node(const VertexId& vertex) {
-    assert(is_vertex_id_valid(vertex) && "Vertex id is not valid!");
-    assert(!vertex_exists_in_graph(vertex, vertices_) &&
+  Graph(VertexDepth max_depth, int new_vertices_num)
+      : max_depth_(max_depth), new_vertices_num_(new_vertices_num) {}
+
+  VertexId source_of_vertex(const VertexId& vertex) {
+    return vertices_[vertex].source_id;
+  }
+
+  void insert_vertex(const VertexId& source_vertex,
+                     const VertexId& vertex,
+                     const VertexDepth& depth) {
+    assert(validation::is_vertex_id_valid(vertex) && "Vertex id is not valid!");
+    assert(!validation::vertex_exists_in_graph(vertex, vertices_) &&
            "Vertex already exists!");
-    vertices_.emplace_back(vertex);
+    vertices_.emplace_back(source_vertex, vertex, depth);
   }
   void insert_edge(const VertexId& source,
                    const VertexId& destination,
-                   const EdgeId& id) {
-    assert(is_edge_id_valid(id) && "Edge id is not valid!");
-    assert(!edge_exists_in_graph(id, edges_) && "Edge already exists!");
-    assert(!are_vertices_connected(source, destination) &&
-           "Vertices are already connected!");
-    assert(vertex_exists_in_graph(source, vertices_) &&
+                   const EdgeId& id,
+                   const Edge::Color& color) {

```

Comment:
nit: Так как по умолчанию у нас создаются серые грани, а все остальные - это особенные случаи, тогда будет логичным указать дефолтное значение этого аргумента:
```suggestion
                   const Edge::Color& color = Edge::Color::Gray) {
```